language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -276,146 +351,92 @@ computeCek con t@(Var _ varName)   = do
-returnCek (FrameUnwrap : con) dat = case dat of
-    IWrap _ _ _ term -> returnCek con term
-    term             ->
-        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just term
+    => Context uni -> CekVal uni -> CekM uni (Plain Term uni)
+--- Instantiate all the free variable of the resulting term in case there are any.
+-- . ◅ V           ↦  [] V
+returnCek [] val = pure $ void $ dischargeCekVal val
+-- s , {_ A} ◅ abs α M  ↦  s ; ρ ▻ M [ α / A ]*
+returnCek (FrameTyInstArg ty : ctx) fun = instantiateEvaluate ctx ty fun
+-- s , [_ (M,ρ)] ◅ V  ↦  s , [V _] ; ρ ▻ M
+returnCek (FrameApplyArg argVarEnv arg : ctx) fun = do
+    -- funVarEnv <- getEnv
+    withEnv argVarEnv $ computeCek (FrameApplyFun fun : ctx) arg
+-- s , [(lam x (M,ρ)) _] ◅ V  ↦  s ; ρ [ x  ↦  V ] ▻ M
+returnCek (FrameApplyFun fun : ctx) arg = do
+    applyEvaluate ctx fun arg
+-- s , wrap A B _ ◅ V  ↦  s ◅ wrap A B V
+returnCek (FrameIWrap ex pat arg : ctx) val =
+    returnCek ctx $ VIWrap ex pat arg val
+-- s , unwrap _ ◅ wrap A B V  ↦  s ◅ V
+returnCek (FrameUnwrap : ctx) val =
+    case val of
+      VIWrap _ _ _ v -> returnCek ctx v
+      _              ->
+        throwingWithCause _MachineError NonWrapUnwrappedMachineError $ Just val
 
 -- | Instantiate a term with a type and proceed.
--- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
--- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
--- apply the term to the type via 'TyInst'.
+-- In case of 'VTyAbs' just ignore the type; for 'VBuiltin', extend
+-- the type arguments with the type, decrement the argument count,
+-- and proceed; otherwise, it's an error.
 instantiateEvaluate
     :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
-    => Context uni -> Type TyName uni ExMemory -> WithMemory Term uni -> CekM uni (Plain Term uni)
-instantiateEvaluate con _ (TyAbs _ _ _ body) = computeCek con body
-instantiateEvaluate con ty fun
-    | isJust $ termAsPrimIterApp fun = returnCek con $ TyInst (memoryUsage () <> memoryUsage fun <> memoryUsage ty) fun ty
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just fun
-
-{- Note [Saved mapping example]
-Consider a polymorphic built-in function @id@, whose type signature on the Plutus side is
-
-    id : all a. a -> a
-
-Notation:
-
-- the variable environment is denoted as @{ <var> :-> (<env>, <value>), ... }@
-- the context is denoted as a Haskell list of 'Frame's
-- 'computeCek' is denoted as @(|>)@
-- 'returnCek' is denoted as @(<|)@
-
-When evaluating
+    => Context uni -> Type TyName uni ExMemory -> CekVal uni -> CekM uni (Plain Term uni)
+instantiateEvaluate ctx _ (VTyAbs _ _ _ body env) = withEnv env $ computeCek ctx body -- FIXME: env?
+instantiateEvaluate ctx ty (VBuiltin ex argEnv bn count tyargs args) =
+    returnCek ctx $ VBuiltin ex argEnv bn (count-1) (ty:tyargs) args
```

Comment:
Do we need an error case for count < 0? Seems worth a note at least.

Reply:
Also, do we need to do something in case this is the last argument?

Reply:
Looking at the code, I think I can get a negative count by overapplying with type arguments, which seems bad.

Reply:
Also can we call this `remainingArgs` or something?

Reply:
> Do we need an error case for count < 0?

Agreed, we should check and throw an error if the counter goes below 0.

> Also, do we need to do something in case this is the last argument?

I think we're just not sure yet what we're going to do regarding interleaved type/term arguments.